subsys/power/device_pm.c

```diff
@@ -0,0 +1,301 @@
+#define DEVICE_PM_RET_SUCCESS		0
+#define DEVICE_PM_RET_QUEUED		1
+
+/* device PM worker queue thread info */
+K_THREAD_STACK_DEFINE(devpm_workq_stack, CONFIG_DEVPM_WORKQUEUE_STACK_SIZE);
+struct k_work_q devpm_workq;
+
+static int set_dev_pm_state(struct device *dev, u32_t request)
+{
+	int ret;
+
+	/* Incase of suspend request, make sure the device is not busy */
+	if (device_busy_check(dev) && (request == DEVICE_PM_SUSPEND_STATE)) {
+		ret = -EBUSY;
+		goto pm_out;
+	}
+
+	ret = device_set_power_state(dev, request);
+
+pm_out:
+	switch (request) {
+	case DEVICE_PM_ACTIVE_STATE:
+		if (ret < 0) {
+			dev->pm->pm_status = DEVICE_PM_STATUS_WAKEUP_ERR;
+		} else {
+			dev->pm->pm_status = DEVICE_PM_STATUS_ACTIVE;
+		}
+		break;
+	case DEVICE_PM_SUSPEND_STATE:
+		if (ret < 0) {
+			dev->pm->pm_status = DEVICE_PM_STATUS_SUSPEND_ERR;
+		} else {
+			dev->pm->pm_status = DEVICE_PM_STATUS_SUSPEND;
+		}
+		break;
+	default:
+		LOG_ERR("Invalid PM state requested (%d)\n", request);
+		return -EINVAL;
+	}
+
+	/* Notify the driver through signal */
+	if (dev->pm->signal) {
+		k_poll_signal_raise(dev->pm->signal, dev->pm->pm_status);
+	}
+
+	return ret;
+}
+
+/*
+ * Device PM worker to handle the Async PM requests
+ * from device_pm_get/device_pm_put API's. The worker
+ * will call the device_set_power_state() to put the
+ * device in desired power state and notify the caller
+ * through signal mechanism.
+ */
+static void pm_work_handler(struct k_work *work)
+{
+	struct device_pm *pm = CONTAINER_OF(work,
+					struct device_pm, dwork);
+	struct device *dev = pm->dev;
+	int ret;
+	u32_t request;
+
+	/* Set async status to in progress */
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	request = dev->pm->async_request;
+	dev->pm->async_req_status =  DEVPM_ASYNC_REQ_INPROG;
+	k_sem_give(&dev->pm->lock);
+
+	/* Certain devices may take lot of time to switch
+	 * between power states. So to avoid blocking of
+	 * Async API's for longer periods we are using a
+	 * dedicated lock to control power state transitins
+	 * between Sync and Async calls.
+	 */
+	k_sem_take(&dev->pm->pm_lock, K_FOREVER);
+	ret = set_dev_pm_state(dev, request);
+	if (ret) {
+		LOG_ERR("Set power state(%d) error(%d)\n", request, ret);
+	}
+	k_sem_give(&dev->pm->pm_lock);
+
+	/* Set async status to done */
+	k_sem_take(&dev->pm->lock, K_FOREVER);
+	dev->pm->async_req_status =  DEVPM_ASYNC_REQ_DONE;
+	k_sem_give(&dev->pm->lock);
+}
+
+/*
+ * Set the device PM state to target PM state based on
+ * the current state. If the PM request is of Async type
+ * then a work item will be submitted to PM worker queue
+ * else it will call device_set_power_state() to set the
+ * device PM state to target state.
+ */
+static int device_pm_state_handler(struct device *dev,
+				   u32_t target_state, u32_t pm_flags)
+{
+	u32_t current_state;
+	int ret = DEVICE_PM_RET_SUCCESS;
+
+	/* Get current device power state */
+	ret = device_get_power_state(dev, &current_state);
```

Comment:
What this API return when device is changing power state now? (note, that power state change is not atomic operation).

Reply:
i think we need to make this atomic or control the flow with lock.